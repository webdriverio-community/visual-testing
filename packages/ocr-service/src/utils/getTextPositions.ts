import getData from './getData.js'
import { getDprPositions } from './index.js'
import type { Line, GetTextPositions, GetTextPositionsOptions } from '../types.js'

export default async function getTextPositions(browser: WebdriverIO.Browser, options: GetTextPositionsOptions): Promise<GetTextPositions[]> {
    const { dpr, filePath, words } = await getData(browser, options)

    return (
        words
            .map(({ text, bbox }:Line) => ({
                dprPosition: getDprPositions(JSON.parse(JSON.stringify(bbox)), dpr),
                filePath,
                originalPosition: bbox,
                text: text.replace(/(^\s+|\s+$)/g, ''),
            }))
            .filter((word) => word.text)
    )
}
