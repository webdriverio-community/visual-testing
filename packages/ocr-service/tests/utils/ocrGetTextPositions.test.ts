import { describe, it, expect, vi, beforeEach } from 'vitest'
import getTextPositions from '../../src/utils/getTextPositions.js'
import getData from '../../src/utils/getData.js'
import { getDprPositions } from '../../src/utils/index.js'

vi.mock('../../src/utils/getData.js', () => ({
    default: vi.fn()
}))
vi.mock('../../src/utils/index.js', () => ({
    getDprPositions: vi.fn().mockImplementation((bbox, dpr) => ({
        left: bbox.left / dpr,
        top: bbox.top / dpr,
        right: bbox.right / dpr,
        bottom: bbox.bottom / dpr
    }))
}))
const options = {
    contrast: 0.25,
    isTesseractAvailable: false,
    language: 'eng',
    ocrImagesPath: 'sample/path/to/',
}

describe('getTextPositions', () => {
    beforeEach(() => {
        vi.clearAllMocks()
    })

    it('returns empty array when no words are present', async () => {
        vi.mocked(getData).mockResolvedValue({
            dpr: 2,
            filePath: 'sample/path/to/image.png',
            text: 'example ',
            lines: [],
            words: []
        })
        const result = await getTextPositions(options)

        expect(result).toEqual([])
        expect(getData).toHaveBeenCalled()
    })

    it('processes words and adjusts their positions based on DPR', async () => {
        vi.mocked(getData).mockResolvedValue({
            dpr: 2,
            filePath: 'sample/path/to/image.png',
            text: 'example ',
            lines: [{ text: 'Sample OCR text', bbox: { left: 1, right: 1, top: 1, bottom: 1 } }],
            words: [
                { text: 'Sample', bbox: { left: 1, right: 1, top: 1, bottom: 1 }, wc: 1 },
                { text: 'OCR t', bbox: { left: 2, right: 2, top: 2, bottom: 2 }, wc: 2 },
                { text: ' ext', bbox: { left: 3, right: 3, top: 3, bottom: 3 }, wc: 3 },
            ]
        })

        const result = await getTextPositions(options)

        expect(getDprPositions).toHaveBeenCalledTimes(3)
        expect(result.length).toBe(3)
        expect(result).toMatchSnapshot()
    })
})
