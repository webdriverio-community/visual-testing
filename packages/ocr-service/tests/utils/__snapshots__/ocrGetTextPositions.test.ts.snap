// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ocrGetTextPositions > processes words and adjusts their positions based on DPR 1`] = `
[
  {
    "dprPosition": {
      "bottom": 0.5,
      "left": 0.5,
      "right": 0.5,
      "top": 0.5,
    },
    "filePath": "sample/path/to/image.png",
    "originalPosition": {
      "bottom": 1,
      "left": 1,
      "right": 1,
      "top": 1,
    },
    "text": "Sample",
  },
  {
    "dprPosition": {
      "bottom": 1,
      "left": 1,
      "right": 1,
      "top": 1,
    },
    "filePath": "sample/path/to/image.png",
    "originalPosition": {
      "bottom": 2,
      "left": 2,
      "right": 2,
      "top": 2,
    },
    "text": "OCR t",
  },
  {
    "dprPosition": {
      "bottom": 1.5,
      "left": 1.5,
      "right": 1.5,
      "top": 1.5,
    },
    "filePath": "sample/path/to/image.png",
    "originalPosition": {
      "bottom": 3,
      "left": 3,
      "right": 3,
      "top": 3,
    },
    "text": "ext",
  },
]
`;
